File: ios/DrawingEngine/Sources/DrawingEngine/Model/Stroke.swift
import Foundation

/// A single freehand stroke captured from the canvas: an ordered list of points.
public struct Stroke: Sendable, Codable, Equatable {
    public var points: [Point2D]

    public init(points: [Point2D]) {
        self.points = points
    }

    public var isDrawable: Bool { points.count >= 2 }

    public var length: Double { Geometry.polylineLength(points) }

    public var boundingBox: BoundingBox? { BoundingBox.of(points) }

    public var start: Point2D? { points.first }
    public var end: Point2D? { points.last }

    /// Gap between the first and last point, relative to the stroke's length.
    /// A small value means the child closed the loop cleanly.
    public var closureGapFraction: Double {
        guard let s = start, let e = end, length > 0 else { return 1 }
        return s.distance(to: e) / length
    }
}

File: web/src/lib/engine/stroke.ts
import type { BoundingBox, Point2D } from './geometry'
import { boundingBox, distance, polylineLength } from './geometry'

export interface Stroke {
  readonly points: readonly Point2D[]
}

export const stroke = (points: readonly Point2D[]): Stroke => ({ points })

export const isDrawable = (s: Stroke): boolean => s.points.length >= 2

export const strokeLength = (s: Stroke): number => polylineLength(s.points)

export const strokeBoundingBox = (s: Stroke): BoundingBox | null => boundingBox(s.points)

/** Gap between first and last as a fraction of length. Small = clean closure. */
export function closureGapFraction(s: Stroke): number {
  if (s.points.length < 2) return 1
  const len = strokeLength(s)
  if (len === 0) return 1
  return distance(s.points[0], s.points[s.points.length - 1]) / len
}
